feat: loop for while lint implementation working

coxmars opened the pull request:
## Summary 📌

- This PR closes issue #34 

This PR introduces a new lint `loop_for_while` that detects and suggests replacing unnecessary `loop` constructs with `while` loops, improving code clarity and conciseness.

## Changes Made ⚡️

### Implementation of `loop_for_while` Lint
- Added a new lint in **`plugin.rs`** to detect `loop` constructs that can be replaced by `while` loops.
- Implemented the fix logic in **`fix.rs`**, ensuring that the conversion from `loop` to `while` preserves the original logic while simplifying the code structure.
- Added diagnostic and fix logic in **`loop_for_while.rs`** to handle conditions and loop bodies appropriately.

### Test Coverage 🥷🏼
- Added test cases in **`test_files/loop_for_while`** to validate the `loop_for_while` lint:
  - Conversion of a simple `loop` with a comparison condition to a `while`.
  - Handling of `loop` with multiple conditions involving logical operators.
  - Conversion of a `loop` with arithmetic conditions.
  - Ensured that the `while` loop correctly reflects the logic of the original `loop`.

### Test Results ✅

All tests have passed successfully, confirming that the `loop_for_while` lint works as intended and improves the readability of the code.

<img width="815" alt="image" src="https://github.com/user-attachments/assets/a16b162a-f8a0-4c02-aec7-c166f7b7bfb6">

0xLucqs commented on crates/cairo-lint-core/src/lints/loop_for_while.rs:
```diff
@@ -0,0 +1,51 @@
+        {
+            let if_block = if_expr.if_block(db);
+
+            let has_unnecessary_parens_break = if_block.statements(db).elements(db).iter().any(|inner_statement| {
+                if let Statement::Break(_) = inner_statement {
+                    inner_statement.as_syntax_node().get_text_without_trivia(db).contains("break ()")
+                } else {
+                    false
+                }
+            });
+
+            if has_unnecessary_parens_break {
+                continue;
+            }
```
```suggestion
```

0xLucqs commented on crates/cairo-lint-core/src/lints/loop_for_while.rs:
```diff
@@ -0,0 +1,51 @@
+                continue;
+            }
+
+            if if_block
+                .statements(db)
+                .elements(db)
+                .iter()
+                .any(|inner_statement| matches!(inner_statement, Statement::Break(_)))
+            {
+                has_break = true;
+                break;
+            }
+        }
+    }
+
+    if has_break {
+        diagnostics.push(PluginDiagnostic {
+            stable_ptr: loop_expr.stable_ptr().untyped(),
+            message: LOOP_FOR_WHILE.to_string(),
+            severity: Severity::Warning,
+        });
+    }
```
```suggestion
            if if_block
                .statements(db)
                .elements(db)
                .iter()
                .any(|inner_statement| matches!(inner_statement, Statement::Break(_)))
            {
                diagnostics.push(PluginDiagnostic {
                    stable_ptr: loop_expr.stable_ptr().untyped(),
                    message: LOOP_FOR_WHILE.to_string(),
                    severity: Severity::Warning,
               });
                return;
            }
        }
    }
```

0xLucqs commented on crates/cairo-lint-core/tests/test_files/breaks/breaks:
```diff
@@ -2,32 +2,32 @@
-        }
-        a += 1;
-    }
+  let mut a = 1_u32;
```
ah yes didn't think about these lints :'( replace the `loop` by `for elt in array![].span()` (it might add another diagnostic but it won't hurt the fix

0xLucqs commented on crates/cairo-lint-core/src/fix.rs:
```diff
@@ -378,4 +379,95 @@ impl Fixer {
+            condition.replace(">=", "<")
+        } else if condition.contains("<=") {
+            condition.replace("<=", ">")
+        } else {
+            format!("!({})", condition)
+        }
```
why not continue this `else if` with `>` `<` `==` `!=` ?

coxmars replied:
> why not continue this `else if` with `>` `<` `==` `!=` ?

Good point! I initially thought that handling the other conditions (like >, <, ==, !=) might not be necessary, but it's important to make the function more comprehensive 🫡

0xLucqs reviewed (commented):
other than that looks good

0xLucqs commented on crates/cairo-lint-core/tests/test_files/loop_for_while/loop_for_while:
```diff
@@ -0,0 +1,181 @@
+//! > loop with arithmetic condition
```
can you add one test like that 
```
fn main() {
    let mut x: u16 = 5;
    loop {
        if x * 2 >= 20 {
            break;
        } else {
            x += 1;
        }
    }
}
```
and one like that:
```
fn main() {
    let mut x: u16 = 5;
    loop {
        if x * 2 >= 20 {
            if x > 30 {
                break;
            }
        } else {
            x += 1;
        }
    }
}
```

coxmars replied:
Yeah sure, I will add them asap

0xLucqs commented on crates/cairo-lint-core/tests/test_files/loop_for_while/loop_for_while:
```diff
@@ -34,6 +34,42 @@ fn main() {
+//! > fixed
+fn main() {
+    let mut x: u16 = 5;
+    while x * 2 < 20 {
+    }
```
this should contain what's in the `else` block
add also a test like that:
```cairo
fn main() {
    let mut x: u16 = 5;
    loop {
        if x * 2 >= 20 {
            break;
        } else {
            x += 1;
        }
        x += 1;
    }
}
```
and the fix should be something like that:
```cairo
fn main() {
    let mut x: u16 = 5;
    while x * 2 < 20 {
         x+=1;
         x+=1;
    }
}

0xLucqs commented on crates/cairo-lint-core/tests/test_files/loop_for_while/loop_for_while:
```diff
@@ -139,13 +175,36 @@ warning: Plugin diagnostic: you seem to be trying to use `loop`. Consider replac
 
 //! > ==========================================================================
 
+//! > loop with nested condition inside if block
```
Nested is a special word that will not trigger the fix. Only use it if you have multiple diagnostics on the same lines for example if you have 2 nested while instead of loop (not here)
```suggestion
//! > loop with multiple condition inside if block
```

coxmars commented:
Hi @0xLucqs, I made some updates. I refactored the code to better handle else clauses and ensure logic both inside and outside the if-else is captured properly. I also added two new tests so let me know if it looks good! 🫡

0xLucqs commented on crates/cairo-lint-core/src/fix.rs:
```diff
@@ -378,4 +379,135 @@ impl Fixer {
+            }
+        }
+        else_body
+    }
+
+    fn invert_condition(condition: &str) -> String {
+        if condition.contains("&&") {
+            condition
+                .split("&&")
+                .map(|part| Self::invert_simple_condition(part.trim()))
+                .collect::<Vec<_>>()
+                .join(" || ")
+        } else if condition.contains("||") {
+            condition
+                .split("||")
+                .map(|part| Self::invert_simple_condition(part.trim()))
+                .collect::<Vec<_>>()
+                .join(" && ")
+        } else {
+            Self::invert_simple_condition(condition)
+        }
+    }
+
+    fn invert_simple_condition(condition: &str) -> String {
+        if condition.contains(">=") {
+            condition.replace(">=", "<")
+        } else if condition.contains("<=") {
+            condition.replace("<=", ">")
+        } else if condition.contains(">") {
+            condition.replace(">", "<=")
+        } else if condition.contains("<") {
+            condition.replace("<", ">=")
+        } else if condition.contains("==") {
+            condition.replace("==", "!=")
+        } else if condition.contains("!=") {
+            condition.replace("!=", "==")
+        } else {
+            format!("!({})", condition)
+        }
+    }
```
can you create a helper module in the fix module? all those functions will for sure be useful to someone else. Also can you check if there is any overlap with the function `indent_snippet` ?

0xLucqs commented on crates/cairo-lint-core/src/fix/helper.rs:
```diff
@@ -0,0 +1,127 @@
+/// 
+/// # Returns
+/// 
+/// A string representing the formatted content of the block.
```
can you add an example of what it does so it's crystal clear ? and same for every function
